lib/sparse/src/index/search_context.rs

```diff
@@ -0,0 +1,454 @@
+        let mut postings_iterators = Vec::new();
+
+        for (query_weight_offset, id) in query.indices.iter().enumerate() {
+            if let Some(posting) = inverted_index.get(id) {
+                postings_iterators.push(IndexedPostingListIterator {
+                    posting_list_iterator: PostingListIterator::new(posting),
+                    query_weight_offset,
+                });
+            }
+        }
+        let result_queue = FixedLengthPriorityQueue::new(top);
+
+        SearchContext {
+            postings_iterators,
+            query,
+            top,
+            result_queue,
+        }
+    }
+
+    /// Example
+    ///
+    /// postings_iterators:
+    ///
+    /// 1,  30, 34, 60, 230
+    /// 10, 30, 35, 51, 230
+    /// 2,  21, 34, 60, 200
+    /// 2,  30, 34, 60, 230
+    ///
+    /// Next:
+    ///
+    /// a,  30, 34, 60, 230
+    /// 10, 30, 35, 51, 230
+    /// 2,  21, 34, 60, 200
+    /// 2,  30, 34, 60, 230
+    ///
+    /// Next:
+    ///
+    /// a,  30, 34, 60, 230
+    /// 10, 30, 35, 51, 230
+    /// b,  21, 34, 60, 200
+    /// b,  30, 34, 60, 230
+    ///
+    /// Next:
+    ///
+    /// a,  30, 34, 60, 230
+    /// c,  30, 35, 51, 230
+    /// b,  21, 34, 60, 200
+    /// b,  30, 34, 60, 230
+    fn advance(&mut self) -> Option<ScoredCandidate> {
+        let min_record_id = Self::next_min(&self.postings_iterators)?;
+        let mut score = 0.0;
+
+        // Iterate second time to advance posting iterators
+        for posting_iterator in self.postings_iterators.iter_mut() {
+            if let Some(record_id) = posting_iterator
+                .posting_list_iterator
+                .peek()
+                .map(|element| element.id)
+            {
+                // accumulate score for the current record id
+                if record_id == min_record_id {
+                    let element = posting_iterator.posting_list_iterator.next().unwrap();
+                    score +=
+                        element.weight * self.query.weights[posting_iterator.query_weight_offset];
+                }
+            }
+        }
+
+        Some(ScoredCandidate {
+            score,
+            vector_id: min_record_id,
+        })
+    }
+
+    pub fn next_min(to_inspect: &[IndexedPostingListIterator<'_>]) -> Option<u32> {
+        // Initialize min record id with max value
+        let mut min_record_id = u32::MAX;
+        // Indicates that posting iterators are not empty
+        let mut found = false;
+
+        // Iterate first time to find min record id at the head of the posting lists
+        for posting_iterator in to_inspect.iter() {
+            if let Some(element) = posting_iterator.posting_list_iterator.peek() {
+                found = true;
+                if element.id < min_record_id {
+                    min_record_id = element.id;
+                }
+            }
+        }
+
+        if !found {
+            return None;
+        }
+        Some(min_record_id)
+    }
```

Comment:
Maybe we better keep track of the minimum ID in an `Option`, so we can simply return that and remove the `found` flag.

Reply:
I need to think about it.
We would have to unpack the Option at each iteration :thinking:

Reply:
Would `to_inspect.iter().filter_map(|pi| pi.posting_list_iterator.peek()).map(|e| e.id).min()` work?

Reply:
Can we, please, not? 😭🙏

I'm too stupid to read tightly packed sequences of iterator adaptors.

We can use and `Option`'al `min_record_id` and simply do:
- `if element.id < min_record_id.unwrap_or(u32::MAX)`
- or `if Some(element.id) < min_record_id || min_record_id.is_none()`

Reply:
I formatted the sequence to have it less tightly packed. Hope that makes it more friendly. 😄

If you read each iterator adapter in sequence, it makes sense: We 1. get all the elements where peek() returns Some(thing), 2. map them to their id and 3. take the `min`imum.

Reply:
I agree that the boolean flag is not idiomatic.

I went with @ffuugoo suggestion instead `if Some(element.id) < min_record_id || min_record_id.is_none()`